File: ios/SsdidWallet/Domain/Profile/ProfileMigration.swift
import Foundation
import SsdidCore

/// One-time migration: copies global profile VC data to the first identity.
enum ProfileMigration {

    private static let legacyProfileId = "urn:ssdid:profile"
    private static let migrationKey = "ssdid_profile_migration_v1"

    static func migrateIfNeeded(vault: Vault) async {
        // Idempotency guard
        guard !UserDefaults.standard.bool(forKey: migrationKey) else { return }

        let credentials = await vault.listCredentials()
        guard let profileVc = credentials.first(where: { $0.id == legacyProfileId }) else {
            // No legacy profile — mark migration complete
            UserDefaults.standard.set(true, forKey: migrationKey)
            return
        }

        let profileName = profileVc.credentialSubject.claims["name"]
        let email = profileVc.credentialSubject.claims["email"]
        let identities = await vault.listIdentities()

        // If no identities, preserve the profile VC so data isn't lost
        guard !identities.isEmpty else { return }

        // Copy profile to the first identity that doesn't already have email
        if let target = identities.first(where: { ($0.email ?? "").isEmpty }) {
            let hasData = !(email ?? "").isEmpty || !(profileName ?? "").isEmpty
            if hasData {
                do {
                    try await vault.updateIdentityProfile(
                        keyId: target.keyId,
                        profileName: profileName,
                        email: email,
                        emailVerified: nil
                    )
                } catch {
                    // Do NOT delete the credential if migration failed — preserve data
                    return
                }
            }
        }

        // Delete the legacy profile VC only after successful migration
        try? await vault.deleteCredential(credentialId: legacyProfileId)
        UserDefaults.standard.set(true, forKey: migrationKey)
    }
}

File: android/app/src/main/java/my/ssdid/wallet/domain/profile/ProfileMigration.kt
package my.ssdid.wallet.domain.profile

import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import my.ssdid.wallet.domain.vault.Vault
import javax.inject.Inject

class ProfileMigration @Inject constructor(private val vault: Vault) {

    private val migrationMutex = Mutex()

    companion object {
        private const val LEGACY_PROFILE_ID = "urn:ssdid:profile"
    }

    suspend fun migrateIfNeeded() = migrationMutex.withLock {
        val profileVc = vault.listCredentials().find { it.id == LEGACY_PROFILE_ID }
            ?: return@withLock

        val profileName = profileVc.credentialSubject.claims["name"]
        val email = profileVc.credentialSubject.claims["email"]
        val identities = vault.listIdentities()

        // If no identities, preserve the profile VC so data isn't lost
        if (identities.isEmpty()) return@withLock

        // Copy profile to the first identity that doesn't already have email
        val target = identities.firstOrNull { it.email.isNullOrBlank() }
        if (target != null && (!email.isNullOrBlank() || !profileName.isNullOrBlank())) {
            vault.updateIdentityProfile(
                target.keyId,
                profileName = profileName,
                email = email
            )
        }

        // Delete the legacy profile VC
        vault.deleteCredential(LEGACY_PROFILE_ID)
    }
}
